### DISCUSSION TITLE
Can a route be both a resource route and a UI route depending on the request?

### DISCUSSION THREAD

**Discussion by lpsinger** [remix-run/remix#6115] [Q&A] [answered]
The Remix documentation explains that a resource route is a route module that does not have a default export.

I would like to create a route that is sometimes a resource route and sometimes a UI route, depending on the MIME type requested by the client through the `Accept:` header. The use case is a route that can provide a UI view in the browser or JSON or plain text when used like a REST API endpoint.

Is this possible in Remix?

**Comment by kiliman** [comment, marked as answer]
Not really, because Remix determines UI and resource routes during the build process.

However, all loaders are technically API endpoints. So if you want to call the loader and just get the JSON response, then you need to pass the route id as the `_data` search param.

```ts
// routes/test.tsx

import { json } from "@remix-run/node";
import { useLoaderData } from "@remix-run/react";

export function loader() {
  return json({ hello: "world" });
}

export default function Test() {
  const data = useLoaderData();
  return <pre>{JSON.stringify(data, null, 2)}</pre>;
}
```

If you call the route `/test`, then you'll get the UI route. However, if you call `/test?_data=routes/test`, you'll just get the JSON response. This is how Remix transitions to CSR. It calls the loaders directly, then renders the UI on the client.

**Reply by kiliman** [reply]
Sorry, fixed typo. The search param is `_data` not `data`.

**Reply by gabimor** [reply]
another typo i belive: it's `routes/test` not `route/test` 
@kiliman thanks for the answer! made things clearer for me,

**Reply by kiliman** [reply]
Ugh.. probably answered before I had my morning coffee.

**Comment by kstoddard** [comment]
@kiliman Do you know if this is documented? Cursory look turned up nothing. Thanks

**Reply by kiliman** [reply]
It's part of the Remix internals, but I don't think it'll change anytime soon.

https://github.com/remix-run/remix/blob/b55b99d6a6daeb7fb1e969370cc59715a0afce80/packages/remix-server-runtime/server.ts#L76-L86

**Reply by kstoddard** [reply]
Thanks for pointing that out!

Yeah, I have a route that started as a resource route but now have a view. I want to be able to continue to hit this route as a resource route for the data but as soon as I add the default export it fails as a resource route. I guess this solution may work but I wonder, do you know of any documented way to do this without worry of this breaking? My first thought is to create a new resource route that both the page with a UI and the other route currently using it can both hit.

**Reply by kiliman** [reply]
Create two routes. One Resource route, one UI route.
```
/myroute.json => return just JSON data
/myroute => returns HTML
```
```ts
// routes/myroute[.json].tsx (Resource Route)
export async function loader({ request }) {
  return json({ message: 'hello world' })
}

// routes/myroute.tsx (UI route)
import { loader } from './myroute[.json]'
export { loader }

export default function Route() {
  const { message } = useLoaderData<typeof loader>()
  return <h1>{ message }</h1>
}
```

**Reply by kstoddard** [reply]
Hmm, how would I do this if I needed to build out the `loader` a bit more in `routes/myroute.tsx`? So for instance, perhaps make other requests etc. Can I make a `loader` to `loader` request?

**Reply by kiliman** [reply]
Since they all run server-side, I'd just extract those as functions that you call from your loaders.
